### ISSUE TITLE
RFC: Mechanism for end2end testing

### ISSUE THREAD

**Issue by mploski** [awslabs/aws-lambda-powertools-python#1226]
### Which AWS Lambda Powertools utility does this relate to?

Other

### Summary

Build mechanism to run end to end tests on Lambda Powertools library using real AWS Services (Lambda, DynamoDB)
Initially tests can be run manually by maintainers on specific branch/commit_id to ensure expected feature works.
Tests should be triggered in GitHub but also maintainers/contributors should be able to run them in their local environment using their own AWS Account.

### Use case

Providing mechanism to run end to end tests in real-live environment allows us to discover different class of problems we cannot find out otherwise by running unit tests ot integration tests with mocks. For example, different behaviour of a library with lambda cold start and with warm start. It also allow us to validate integration with external services (CloudWatch Logs, X-ray) and ensure final real user experience is what we expect. 

### Proposal

Github configuration:

1. Integrate GitHub with AWS Account using OIDC: https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/configuring-openid-connect-in-amazon-web-services
2. Specify end2end tests commands in Github Actions
<img width="571" alt="Screenshot 2022-05-26 at 11 24 50" src="https://user-images.githubusercontent.com/132872/170469705-81817ddb-fe4b-4673-9b44-29e8053bfa5b.png">

Test setup:

1. Separate end to end tests directories by Powertools functionality like logger/tracer/metrics 
<img width="115" alt="Screenshot 2022-05-26 at 11 30 39" src="https://user-images.githubusercontent.com/132872/170469911-54bba47a-e095-4fc6-8057-e3601251fbf9.png">
2. Create pytest fixture in shared conftest file that run infrastructure helper class to deploy and run lambda at the begging of the test session. It uses handler python file developed by user and stored in test `handler` directory. Fixture is triggered for every functionality tests group. At the end of test session fixture destroy CloudFormation stack automatically.
<img width="513" alt="Screenshot 2022-05-26 at 11 32 53" src="https://user-images.githubusercontent.com/132872/170470288-d8b1dccc-4763-4cd9-a560-c5fdca7f6ac8.png">
3. Infrastructure deployment class does the following. 

- deploy method is an entry point
    
    <img width="494" alt="Screenshot 2022-05-26 at 11 43 47" src="https://user-images.githubusercontent.com/132872/170470382-d51497df-d373-47a7-9048-1bf6259e1082.png">

- Find all python files in test handler directory

- Use CDK SDK for python to synthesise python code into CloudFormation stack + generate assets inside `asset` directory.

    <img width="648" alt="Screenshot 2022-05-26 at 11 35 20" src="https://user-images.githubusercontent.com/132872/170470560-0a5fdc09-63b3-4658-8382-53d14e6367f4.png">

- Push assets to s3 bucket specified in CloudFormation stack

    <img width="547" alt="Screenshot 2022-05-26 at 11 36 22" src="https://user-images.githubusercontent.com/132872/170470656-e81807e3-fcfb-4786-a697-ef8b7d741df2.png">

- Deploy stack into aws account

    <img width="576" alt="Screenshot 2022-05-26 at 11 36 52" src="https://user-images.githubusercontent.com/132872/170470692-c2629b94-ddb2-4275-a4bb-b031ae994c43.png">
4. Create tests in features e2e directories

- Add Handlers files into handlers subdir

   <img width="167" alt="Screenshot 2022-05-26 at 11 45 35" src="https://user-images.githubusercontent.com/132872/170471341-b9fd5d86-11c0-431f-b8dd-c2bef825066b.png">

- create test_<feature>.py file. For Example: test_logger.py

- create fixture for lambda configuration that set specific environment variables or parameters

  <img width="449" alt="Screenshot 2022-05-26 at 11 23 09" src="https://user-images.githubusercontent.com/132872/170471508-b9b88b6b-86a1-4cb3-a322-5c4feed29409.png">

- Create helper method allowing us to fetch logs from cloudwatch

  <img width="719" alt="Screenshot 2022-05-26 at 11 49 36" src="https://user-images.githubusercontent.com/132872/170471651-e862a7b8-f232-4242-94bc-a4dc20acb694.png">

- Create test using two fixtures and helper method

  <img width="753" alt="Screenshot 2022-05-26 at 11 51 15" src="https://user-images.githubusercontent.com/132872/170471749-def5b484-59bc-4e62-893d-c9d542e0d346.png">

5. Trigger tests

- From your local environment

  <img width="707" alt="Screenshot 2022-05-26 at 12 16 11" src="https://user-images.githubusercontent.com/132872/170471951-265e8757-acbe-4bb5-b3e6-512f7f6c921e.png">

- From GitHub

  <img width="707" alt="Screenshot 2022-05-26 at 12 17 29" src="https://user-images.githubusercontent.com/132872/170472005-f18b39ab-d007-48ad-8bcf-753261273cf8.png">

### Out of scope

1. Automated tests run on PR opened/comments/labels
2. Extensive set of tests

### Potential challenges

1. By using `pytest xdist` plugin we can easily parallelise tests per group and create infrastructure and run tests in parallel. This leads to Powertools lambda layer being created 3 times, which put a pressure on CPU/RAM/IOPS unnecessarily. We should optimise the solution to create layer only once and then run parallelised tests with reference to this layer.
2. Currently as a prerequisite we  bootstraped account for CDK usage by issuing `cdk boostrap`. Since s3 created by cdk doesn't have lifecycle policy to remove old artefacts we need to customise default template used by CDK bootstrap command and attach it to the solution with good description how to use it.

### Alternative solutions

```markdown
1. Use CDK CLI to deploy infrastructure directly, not custom code to synthesise cdk code, then deploy assets and run AWS CloudFormation deployment - dropped to avoid running CLI from subprocess (more
latency added) + to avoid additional node dependencies
2. Write AWS CodeBuild pipeline on AWS Account side that would run tests stored somewhere else outside of the project. No configuration and tests exposed in powertools repo - dropped due to initial assumption that we want end to end tests to be part of the project/increase visibility/allow contributors to run those tests during development phase on their own
```

### Acknowledgment

- [X] This feature request meets [Lambda Powertools Tenets](https://awslabs.github.io/aws-lambda-powertools-python/latest/#tenets)
- [ ] Should this be considered in other Lambda Powertools languages? i.e. [Java](https://github.com/awslabs/aws-lambda-powertools-java/), [TypeScript](https://github.com/awslabs/aws-lambda-powertools-typescript/)

**Comment by heitorlessa** [comment]
Hey @mploski, thanks a lot for pulling that together. I liked it!! I have some ideas on how we can improve the RFC before you ship the PR you have staged:

* **Focus on mechanics over fine-grained details**. In the `details` section, you could provide a high-level overview then zoom in each area with your rationale without going in fixture line of code as that will likely change. For example, **1/** Share one or two end-to-end tests to demonstrate what a test looks like, **2/** Explain how deployment and teardown will look like, and **3/** Paint the picture of how a folder structure would look like.
* **Share what type of test is best suited for this**. One thing we discussed in the last maintainers sync was using E2E test to cover `Lambda + Event Source configuration + Potential side effects` (Batching, etc) that are troublesome to do and maintain at runtime. For integ test, simpler tests could create resources at runtime for efficiency (Parameters, etc.). This will help us when we look back at this RFC for future understanding, or cross-referencing an upcoming integ test RFC.
* **Share your benchmark**. You've done a lot of work in assessing whether to use mutable functions with CDK vs multiple immutable functions vs SAM CLI vs SDK. You can share that as a GitHub Gist and link to it under `Alternative solutions` - this might help future customers looking to do the same.

These help readers to stay focused on the mechanics and whether they can see any immediate areas for improvement, and worry less about optimizing the implementations early in the process.

A diagram of the workflow could also help, especially now that GitHub has native [Mermaid integration](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/), but don't worry about it as I know bandwidth is limited.

Feel free to ping me on areas you'd like to offload to me too <3

Thanks!

**Comment by mploski** [comment]
> Hey @mploski, thanks a lot for pulling that together. I liked it!! I have some ideas on how we can improve the RFC before you ship the PR you have staged:
> 
>     * **Focus on mechanics over fine-grained details**. In the `details` section, you could provide a high-level overview then zoom in each area with your rationale without going in fixture line of code as that will likely change. For example, **1/** Share one or two end-to-end tests to demonstrate what a test looks like, **2/** Explain how deployment and teardown will look like, and **3/** Paint the picture of how a folder structure would look like.
> 
>     * **Share what type of test is best suited for this**. One thing we discussed in the last maintainers sync was using E2E test to cover `Lambda + Event Source configuration + Potential side effects` (Batching, etc) that are troublesome to do and maintain at runtime. For integ test, simpler tests could create resources at runtime for efficiency (Parameters, etc.). This will help us when we look back at this RFC for future understanding, or cross-referencing an upcoming integ test RFC.
> 
>     * **Share your benchmark**. You've done a lot of work in assessing whether to use mutable functions with CDK vs multiple immutable functions vs SAM CLI vs SDK. You can share that as a GitHub Gist and link to it under `Alternative solutions` - this might help future customers looking to do the same.
> 
> 
> These help readers to stay focused on the mechanics and whether they can see any immediate areas for improvement, and worry less about optimizing the implementations early in the process.
> 
> A diagram of the workflow could also help, especially now that GitHub has native [Mermaid integration](https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/), but don't worry about it as I know bandwidth is limited.
> 
> Feel free to ping me on areas you'd like to offload to me too <3
> 
> Thanks!

Hey @heitorlessa, thanks for your feedback. I restructured this RFC based on your suggestion. Can you take a look one more time?

**Comment by mploski** [comment]
> This is much much better, thank you for taking all this time @mploski - I've made line editing by re-arranging some content, shorten a few, while keeping the same meaning. Please have another look in case I accidentally removed anything critical to the idea.
> 
> I removed the `Triggering Tests` section, moved some of the `Separating Tests from Infrastructure Deployment` section then removed it. They were primarily implementation details and they might change rapidly as the PR starts.
> 
> Please go ahead with the PR @mploski ;-)

Added small adjustments around infrastructure class re-usage. It looks much more concise @heitorlessa, thx! Will issue PR shortly (within this week)

**Comment by mploski** [comment]
@heitorlessa FYI Still working on a PR, found some issues I want to solve first ( too-fat generated powertools layer or AWS API throttling causing tests errors if we run our tests from Github Actions with runtime matrix).

**Comment by alexpulver** [comment]
Hey! Left some comments on the PR itself. Regarding [CF Stack vs CDK Hotswap vs Direct API Hotswap](https://github.com/awslabs/aws-lambda-powertools-python/files/8833059/CF.stack.vs.hotswap.using.pdf) - what about using CDK Hotswap AND multiple Lambda functions?

**Comment by heitorlessa** [comment]
Small update to clarify that we will run E2E tests on every merge to `develop` (trunk). We can run it manually (already implemented) and maintainers will exercise judgement on whether to do it before releasing it. As we add more E2E tests, and find ways to simplify CDK, we can eventually add E2E as a step in the release process - there's a ton of other work to do so we don't need to prioritize it now.

Last thing to test is whether it's worth using `cdk-assets` directly and benchmark it. If it doesn't add any significant overhead we could go ahead with it. If at all possible, we shouldn't rely on CDK CLI and any Node ecosystem as we won't have the bandwidth to deal with the operational and security overhead, as we've been there before.

Thank you so much again @mploski for starting this, going through all comments, and such.

**Comment by mploski** [comment]
> Small update to clarify that we will run E2E tests on every merge to `develop` (trunk). We can run it manually (already implemented) and maintainers will exercise judgement on whether to do it before releasing it. As we add more E2E tests, and find ways to simplify CDK, we can eventually add E2E as a step in the release process - there's a ton of other work to do so we don't need to prioritize it now.
> 
> Last thing to test is whether it's worth using `cdk-assets` directly and benchmark it. If it doesn't add any significant overhead we could go ahead with it. If at all possible, we shouldn't rely on CDK CLI and any Node ecosystem as we won't have the bandwidth to deal with the operational and security overhead, as we've been there before.
> 
> Thank you so much again @mploski for starting this, going through all comments, and such.

Based on the discussion with @alexpulver I analysed the code of cdk-assets npm package and also design for it: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0092-asset-publishing.md. Currently I'm updating the PR to rely on this design for my custom asset generation and later this week I will compare it with npm cdk-assets direct usage for assets upload.

**Comment by heitorlessa** [comment]
This is now merged :) We'll be communicating more details as part of the release. Additional enhancements and E2E tests for other utilities will be dealt separately.

HUGE thank you @mploski for going through this length - intense review process, benchmarking multiple options, documentation, etc - and thanks for all reviewers, truly; it takes a village!

**Comment by heitorlessa** [comment]
Updating here for future correctness, as we moved to CDK CLI due to context methods `from_lookup` only works with the CLI. We also implemented Lambda Layer caching and moved away from Docker.

Pasting the section in the maintainers playbook about the framework.

----

## E2E framework

### Structure

Our E2E framework relies on [Pytest fixtures](https://docs.pytest.org/en/6.2.x/fixture.html) to coordinate infrastructure and test parallelization - see [Test Parallelization](#test-runner-parallelization) and [CDK CLI Parallelization](#cdk-cli-parallelization).

**tests/e2e structure**

```shell
.
├── __init__.py
├── conftest.py # builds Lambda Layer once
├── logger
│   ├── __init__.py
│   ├── conftest.py  # deploys LoggerStack
│   ├── handlers
│   │   └── basic_handler.py
│   ├── infrastructure.py # LoggerStack definition
│   └── test_logger.py
├── metrics
│   ├── __init__.py
│   ├── conftest.py  # deploys MetricsStack
│   ├── handlers
│   │   ├── basic_handler.py
│   │   └── cold_start.py
│   ├── infrastructure.py # MetricsStack definition
│   └── test_metrics.py
├── tracer
│   ├── __init__.py
│   ├── conftest.py  # deploys TracerStack
│   ├── handlers
│   │   ├── async_capture.py
│   │   └── basic_handler.py
│   ├── infrastructure.py  # TracerStack definition
│   └── test_tracer.py
└── utils
    ├── __init__.py
    ├── data_builder  # build_service_name(), build_add_dimensions_input, etc.
    ├── data_fetcher  # get_traces(), get_logs(), get_lambda_response(), etc.
    ├── infrastructure.py # base infrastructure like deploy logic, etc.
```

Where:

- **`<feature>/infrastructure.py`**. Uses CDK to define the infrastructure a given feature needs.
- **`<feature>/handlers/`**. Lambda function handlers to build, deploy, and exposed as stack output in PascalCase (e.g., `BasicHandler`).
- **`utils/`**. Test utilities to build data and fetch AWS data to ease assertion
- **`conftest.py`**. Deploys and deletes a given feature infrastructure. Hierarchy matters:
    - **Top-level (`e2e/conftest`)**. Builds Lambda Layer only once and blocks I/O across all CPU workers.
    - **Feature-level (`e2e/<feature>/conftest`)**. Deploys stacks in parallel and make them independent of each other.

### Mechanics

Under [`BaseInfrastructure`](https://github.com/awslabs/aws-lambda-powertools-python/blob/develop/tests/e2e/utils/infrastructure.py), we hide the complexity of deployment and delete coordination under `deploy`, `delete`, and `create_lambda_functions` methods.

This allows us to benefit from test and deployment parallelization, use IDE step-through debugging for a single test, run one, subset, or all tests and only deploy their related infrastructure, without any custom configuration.

> Class diagram to understand abstraction built when defining a new stack (`LoggerStack`)

```mermaid
classDiagram
    class InfrastructureProvider {
        <<interface>>
        +deploy() Dict
        +delete()
        +create_resources()
        +create_lambda_functions() Dict~Functions~
    }

    class BaseInfrastructure {
        +deploy() Dict
        +delete()
        +create_lambda_functions() Dict~Functions~
        +add_cfn_output()
    }

    class TracerStack {
        +create_resources()
    }

    class LoggerStack {
        +create_resources()
    }

    class MetricsStack {
        +create_resources()
    }

    class EventHandlerStack {
        +create_resources()
    }

    InfrastructureProvider <|-- BaseInfrastructure : implement
    BaseInfrastructure <|-- TracerStack : inherit
    BaseInfrastructure <|-- LoggerStack : inherit
    BaseInfrastructure <|-- MetricsStack : inherit
    BaseInfrastructure <|-- EventHandlerStack : inherit
```

### Authoring a new feature E2E test

Imagine you're going to create E2E for Event Handler feature for the first time. Keep the following mental model when reading:

```mermaid
graph LR
    A["1. Define infrastructure"]-->B["2. Deploy/Delete infrastructure"]-->C["3.Access Stack outputs" ]
```

#### 1. Define infrastructure

We use CDK as our Infrastructure as Code tool of choice. Before you start using CDK, you'd take the following steps:

1. Create `tests/e2e/event_handler/infrastructure.py` file
2. Create a new class `EventHandlerStack` and inherit from `BaseInfrastructure`
3. Override `create_resources` method and define your infrastructure using CDK
4. (Optional) Create a Lambda function under `handlers/alb_handler.py`

> Excerpt `tests/e2e/event_handler/infrastructure.py`

```python
class EventHandlerStack(BaseInfrastructure):
    def create_resources(self):
        functions = self.create_lambda_functions()

        self._create_alb(function=functions["AlbHandler"])
        ...

    def _create_alb(self, function: Function):
        vpc = ec2.Vpc.from_lookup(
            self.stack,
            "VPC",
            is_default=True,
            region=self.region,
        )

        alb = elbv2.ApplicationLoadBalancer(self.stack, "ALB", vpc=vpc, internet_facing=True)
        CfnOutput(self.stack, "ALBDnsName", value=alb.load_balancer_dns_name)
        ...
```

> Excerpt `tests/e2e/event_handler/handlers/alb_handler.py`

```python
from aws_lambda_powertools.event_handler import ALBResolver, Response, content_types

app = ALBResolver()


@app.get("/todos")
def hello():
    return Response(
        status_code=200,
        content_type=content_types.TEXT_PLAIN,
        body="Hello world",
        cookies=["CookieMonster", "MonsterCookie"],
        headers={"Foo": ["bar", "zbr"]},
    )


def lambda_handler(event, context):
    return app.resolve(event, context)
```

#### 2. Deploy/Delete infrastructure when tests run

We need to create a Pytest fixture for our new feature under `tests/e2e/event_handler/conftest.py`.

This will instruct Pytest to deploy our infrastructure when our tests start, and delete it when they complete whether tests are successful or not. Note that this file will not need any modification in the future.

> Excerpt `conftest.py` for Event Handler

```python
import pytest

from tests.e2e.event_handler.infrastructure import EventHandlerStack


@pytest.fixture(autouse=True, scope="module")
def infrastructure():
    """Setup and teardown logic for E2E test infrastructure

    Yields
    ------
    Dict[str, str]
        CloudFormation Outputs from deployed infrastructure
    """
    stack = EventHandlerStack()
    try:
        yield stack.deploy()
    finally:
        stack.delete()

```

#### 3. Access stack outputs for E2E tests

Within our tests, we should now have access to the `infrastructure` fixture we defined earlier in `tests/e2e/event_handler/conftest.py`.

We can access any Stack Output using pytest dependency injection.

> Excerpt `tests/e2e/event_handler/test_header_serializer.py`

```python
@pytest.fixture
def alb_basic_listener_endpoint(infrastructure: dict) -> str:
    dns_name = infrastructure.get("ALBDnsName")
    port = infrastructure.get("ALBBasicListenerPort", "")
    return f"http://{dns_name}:{port}"


def test_alb_headers_serializer(alb_basic_listener_endpoint):
    # GIVEN
    url = f"{alb_basic_listener_endpoint}/todos"
    ...
```

### Internals

#### Test runner parallelization

Besides speed, we parallelize our end-to-end tests to ease asserting async side-effects may take a while per test too, _e.g., traces to become available_.

The following diagram demonstrates the process we take every time you use `make e2e` locally or at CI:

```mermaid
graph TD
    A[make e2e test] -->Spawn{"Split and group tests <br>by feature and CPU"}

    Spawn -->|Worker0| Worker0_Start["Load tests"]
    Spawn -->|Worker1| Worker1_Start["Load tests"]
    Spawn -->|WorkerN| WorkerN_Start["Load tests"]

    Worker0_Start -->|Wait| LambdaLayer["Lambda Layer build"]
    Worker1_Start -->|Wait| LambdaLayer["Lambda Layer build"]
    WorkerN_Start -->|Wait| LambdaLayer["Lambda Layer build"]

    LambdaLayer -->|Worker0| Worker0_Deploy["Launch feature stack"]
    LambdaLayer -->|Worker1| Worker1_Deploy["Launch feature stack"]
    LambdaLayer -->|WorkerN| WorkerN_Deploy["Launch feature stack"]

    Worker0_Deploy -->|Worker0| Worker0_Tests["Run tests"]
    Worker1_Deploy -->|Worker1| Worker1_Tests["Run tests"]
    WorkerN_Deploy -->|WorkerN| WorkerN_Tests["Run tests"]

    Worker0_Tests --> ResultCollection
    Worker1_Tests --> ResultCollection
    WorkerN_Tests --> ResultCollection

    ResultCollection{"Wait for workers<br/>Collect test results"}
    ResultCollection --> TestEnd["Report results"]
    ResultCollection --> DeployEnd["Delete Stacks"]
```

#### CDK CLI parallelization

For CDK CLI to work with [independent CDK Apps](https://docs.aws.amazon.com/cdk/v2/guide/apps.html), we specify an output directory when synthesizing our stack and deploy from said output directory.

```mermaid
flowchart TD
    subgraph "Deploying distinct CDK Apps"
        EventHandlerInfra["Event Handler CDK App"] --> EventHandlerSynth
        TracerInfra["Tracer CDK App"] --> TracerSynth
       EventHandlerSynth["cdk synth --out cdk.out/event_handler"] --> EventHandlerDeploy["cdk deploy --app cdk.out/event_handler"]

       TracerSynth["cdk synth --out cdk.out/tracer"] --> TracerDeploy["cdk deploy --app cdk.out/tracer"]
    end
```

We create the typical CDK `app.py` at runtime when tests run, since we know which feature and Python version we're dealing with (locally or at CI).

> Excerpt `cdk_app_V39.py` for Event Handler created at deploy phase

```python
from tests.e2e.event_handler.infrastructure import EventHandlerStack
stack = EventHandlerStack()
stack.create_resources()
stack.app.synth()
```

When we run E2E tests for a single feature or all of them, our `cdk.out` looks like this:

```shell
total 8
drwxr-xr-x  18 lessa  staff   576B Sep  6 15:38 event-handler
drwxr-xr-x   3 lessa  staff    96B Sep  6 15:08 layer_build
-rw-r--r--   1 lessa  staff    32B Sep  6 15:08 layer_build.diff
drwxr-xr-x  18 lessa  staff   576B Sep  6 15:38 logger
drwxr-xr-x  18 lessa  staff   576B Sep  6 15:38 metrics
drwxr-xr-x  22 lessa  staff   704B Sep  9 10:52 tracer
```

```mermaid
classDiagram
    class CdkOutDirectory {
        feature_name/
        layer_build/
        layer_build.diff
    }

    class EventHandler {
        manifest.json
        stack_outputs.json
        cdk_app_V39.py
        asset.uuid/
        ...
    }

    class StackOutputsJson {
        BasicHandlerArn: str
        ALBDnsName: str
        ...
    }

    CdkOutDirectory <|-- EventHandler : feature_name/
    StackOutputsJson <|-- EventHandler
```

Where:

- **`<feature>`**. Contains CDK Assets, CDK `manifest.json`, our `cdk_app_<PyVersion>.py` and `stack_outputs.json`
- **`layer_build`**. Contains our Lambda Layer source code built once, used by all stacks independently
- **`layer_build.diff`**. Contains a hash on whether our source code has changed to speed up further deployments and E2E tests

Together, all of this allows us to use Pytest like we would for any project, use CDK CLI and its [context methods](https://docs.aws.amazon.com/cdk/v2/guide/context.html#context_methods) (`from_lookup`), and use step-through debugging for a single E2E test without any extra configuration.

> NOTE: VSCode doesn't support debugging processes spawning sub-processes (like CDK CLI does w/ shell and CDK App). Maybe [this works](https://stackoverflow.com/a/65339352). PyCharm works just fine.
